### DISCUSSION TITLE
Freeplane can not use Java version >= 24 because of removal of the Java Security Manager

### DISCUSSION THREAD

**Discussion by dpolivaev** [freeplane/freeplane#2124] [Announcements]
In Java 24 security manager is being removed without any replacement, see https://openjdk.org/jeps/486.
Because having no security manager would allow any groovy script, specifically any formula, to read and modify any files and use network access without restrictions,Freeplane can not use Java 24 or later. It means that the latest Java long term support version used with Freeplane ist Java 21, and Freeplane should refuse running under Java 24 or later.

**Comment by mkalinski** [comment]
Is there a plan to replace the security manager, or will Freeplane require staying on Java 21 indefinitely? There's been no movement on #841 in 2 years.

**Reply by dpolivaev** [reply]
I have not heard of any replacement for the security manager, so I think Freeplane can only use Java 21 LTS.

**Reply by mkalinski** [reply]
Java 21 LTS is supposed to be supported for many more years, for sure, but when it reaches EOL, I think if the plan is to make users install an outdated Java version to continue using Freeplane, then it's a bigger security risk than the absence of security manager.

And from what I can read, Java security manager was deprecated for being really old and not well designed, so it's unclear to me if it really protects Freeplane users if someone was determined to execute malicious code using groovy scripts.

**Reply by dpolivaev** [reply]
You do not need to expose Java version used with Freeplane to any other applications. Therefore I do not think that security is likely to be compromised even when Java stops getting updates. And at the moment all known LTS versions starting with Java 8 are still getting updates, so there is a hope that it continues.

**Reply by DestyNova** [reply]
It also seems to fail with Java 21.0.2 (OpenJDK). It seems a much older version may be needed (17?).

**Reply by logrusx** [reply]
> It also seems to fail with Java 21.0.2 (OpenJDK). It seems a much older version may be needed (17?).

Works fine with OpenJDK 21.0.6 here.

**Reply by riasevern** [reply]
Which OS are you using logrusx?
I ask because I'm on mint 22.1 and I can't get Freeplane to run with java version 21 no matter how I try to install the JRE or JDK. I've dropped back to 17 as a workaround. In searching for a cause/fix I came across this:
       https://forums.linuxmint.com/viewtopic.php?t=425994
I'm not knowledgeable enough to understand whether this is a distro/repository issue or a Java issue but it seems like for some there is an issue.

**Reply by dpolivaev** [reply]
Post here your console output

**Reply by logrusx** [reply]
@riasevern, I'm using Gentoo which doesn't mean much. Gentoo is a rolling distribution, it changes every minute.

Regarding the link you've posted, I'm not familiar with Mint terminology.

It's best to start freeplane from the console and post the output.

**Reply by riasevern** [reply]
I'm running Linux Mint 22.1 Xia

Had to recreate so:_

- Removed Freeplane & all versions of java on system
- Installed freeplane using software manager. This installs 1.7.10.1. According to s/w mgr.  dependencies include java 21 which should be installed. Indeed there is a directory created: /usr/lib/jvm/java-21-openjdk-amd64 with lots of bits.
- Set FREEPLANE_JAVA_HOME to /usr/lib/jvm/java-21-openjdk-amd64/ in .profile.


From next login:

Terminal session as follows (ignore the strikethrough, not sure what caused that in the markup):

rick@Alvega:~$ echo $FREEPLANE_JAVA_HOME
/usr/lib/jvm/java-21-openjdk-amd64/
rick@Alvega:~$ freeplane
[warning] /usr/bin/freeplane: No java runtime was found
[error] /usr/bin/freeplane: Unable to find an appropriate java runtime. See java_wrappers(7) for help
rick@Alvega:~$ java -version
openjdk version "21.0.7" 2025-04-15
OpenJDK Runtime Environment (build 21.0.7+6-Ubuntu-0ubuntu124.04)
OpenJDK 64-Bit Server VM (build 21.0.7+6-Ubuntu-0ubuntu124.04, mixed mode, sharing)


Use s/w mgr to install openjdk-17-jre (version shown in s/w mgr is 17.0.15+6~us1-0ubuntu1~24.04)


now immediately in terminal I see the following:

rick@Alvega:~$ java -version
openjdk version "21.0.7" 2025-04-15
OpenJDK Runtime Environment (build 21.0.7+6-Ubuntu-0ubuntu124.04)
OpenJDK 64-Bit Server VM (build 21.0.7+6-Ubuntu-0ubuntu124.04, mixed mode, sharing)

This is still reporting 21 installed presumably both are there
So try to run freeplane from command line


rick@Alvega:~$ freeplane
*org.knopflerfish.framework.readonly=true
*org.knopflerfish.gosg.jars=reference:file:/usr/share/freeplane/core/
*org.freeplane.user.dir=/home/rick


....... rest of o/p attached as text

and freeplane fires up on the screen. Bit old, but that's repositories.

Whether freeplane is actually running using the 17 virtual machine or the or 21 version I have no idea, that's beyond me I'm afraid.

Happy to provide more info if you can give me some step by step instructions.

[terminal output.txt](https://github.com/user-attachments/files/20690801/terminal.output.txt)

**Reply by macmarrum** [reply]
Hi @riasevern 

In your [terminal output.txt](https://github.com/user-attachments/files/20690801/terminal.output.txt), I can see the following

```
INFO: freeplane_version = 1.7.10; freeplane_xml_version = freeplane 1.7.0
git revision = 34e2242e73d5cf0190a189ef0a60e95b983a3bf1 Dimitry Polivaev 2019-10-25T18:36:32Z
java_version = 17.0.15; os_name = Linux; os_version = 6.8.0-60-generic
```

It looks like you are running Freeplane 1.7.10. This version is from 2019. I'm not sure it supports running on recent Java versions.

You can try to download and install the latest **.deb** (`freeplane_1.*.*~upstream-1_all.deb`) from https://sourceforge.net/projects/freeplane/files/freeplane%20stable/

**Comment by wordmuse** [comment]
I noticed that in the SourceForge download folder, _Freeplane-Setup-with-Java-1.12.7-pre08.exe_ is the most recent with Java. Going forward from that, I can use the most recent .exe download, even though it doesn't have the Java element as part of it?

~ Bal

**Comment by dpolivaev** [comment]
I am planing to support Java range from 17 to 25 instead of 8 to 21 sometimes later.
It will not happen over night.

If you are curious and want to review my plans you can read https://github.com/freeplane/freeplane/blob/1.13.x/ai-specs/tasks/backlog/replace-script-formula-sandbox-for-java-25.md and comment here.

**Reply by zmarcos** [reply]
There is no mention of protection from **Java Serialization**. 
What if some malicious script overrides serialization filters.

Was the use of any GraalVM asset ever considered?
https://www.graalvm.org/latest/security-guide/sandboxing/
https://www.graalvm.org/latest/reference-manual/espresso/

If yes, why a custom implementation is better than using the sandboxing offered by the GraalVM stack (Polyglot/Truffle/Espresso)?

My recommendation for the plan:
***Anything that avoids the use of Java agents is worth of consideration. If possible, explore any method to make the **script compiler** replace(or remove) anything unsafe in the Java bytecode.***

**Reply by dpolivaev** [reply]
@zmarcos thank you for your points, I updated the task to consider them. Shortly, because currently freeplane maps use groovy scripts operating on Java objects directly, the sandboxing in the GraalVM way does not seem to be a good solution. Serialization protection is considered now.
